Kernel/Bus/USB/USBPipe.cpp

```diff
@@ -71,4 +71,31 @@ ErrorOr<size_t> Pipe::control_transfer(u8 request_type, u8 request, u16 value, u
     return transfer_length;
 }
 
+ErrorOr<size_t> Pipe::bulk_transfer_in(u16 length, void* data)
+{
+    auto transfer = TRY(Transfer::try_create(*this, length));
+
+    dbgln_if(USB_DEBUG, "Pipe: Bulk in transfer allocated @ {}", transfer->buffer_physical());
+
+    auto transfer_length = TRY(m_controller->submit_bulk_transfer(*transfer));
+    memcpy(data, transfer->buffer().as_ptr(), transfer_length);
+
+    dbgln_if(USB_DEBUG, "Pipe: Bulk in transfer complete!");
+
+    return transfer_length;
+}
+
+ErrorOr<size_t> Pipe::bulk_transfer_out(u16 length, void* data)
+{
+    auto transfer = TRY(Transfer::try_create(*this, length));
+    transfer->write_buffer(length, data);
+
+    dbgln_if(USB_DEBUG, "Pipe: Bulk out transfer allocated @ {}", transfer->buffer_physical());
+
+    auto transfer_length = TRY(m_controller->submit_bulk_transfer(*transfer));
+    dbgln_if(USB_DEBUG, "Pipe: Bulk out transfer complete!");
+
+    return transfer_length;
+}
+
```

Comment:
Is there any particular reason these are separate functions? A pipe's direction is specified by the member `m_direction`. It seems the only difference is the debug message and call to `memcpy`? In that case, can we merge these into one function and you just switch on the value of `m_direction` to print the debug log?

Reply:
Agreed, added a commit to fix this.

Reply:
@Quaker762 also `transfer->write_buffer(length, data)` is happening in only one of the two functions.

I feel the differences and semantics warrant two separate functions, which also improves the readability of the calling code.